lsp-types/src/Language/LSP/Types/Uri.hs

```diff
@@ -151,37 +184,78 @@ platformAdjustToUriPath systemOS srcPath
           FPP.addTrailingPathSeparator (init drv)
       | otherwise = drv
 
+#ifdef OS_PATH
+type OsPath = OsPath.OsPath
+#else
+type OsPath = FilePath
+#endif
+
 -- | Newtype wrapper around FilePath that always has normalized slashes.
 -- The NormalizedUri and hash of the FilePath are cached to avoided
 -- repeated normalisation when we need to compute them (which is a lot).
 --
 -- This is one of the most performance critical parts of ghcide, do not
 -- modify it without profiling.
-data NormalizedFilePath = NormalizedFilePath NormalizedUri !FilePath
+data NormalizedFilePath = NormalizedFilePath !NormalizedUri {-# UNPACK #-} !ShortByteString
     deriving (Generic, Eq, Ord)
 
 instance NFData NormalizedFilePath
 
 instance Binary NormalizedFilePath where
   put (NormalizedFilePath _ fp) = put fp
   get = do
-    v <- Data.Binary.get :: Get FilePath
-    let nuri = internalNormalizedFilePathToUri v
-    return (normalizedFilePath nuri v)
-
--- | A smart constructor that performs UTF-8 encoding and hash consing
-normalizedFilePath :: NormalizedUri -> FilePath -> NormalizedFilePath
-normalizedFilePath nuri nfp = NormalizedFilePath nuri nfp
+    v <- Data.Binary.get :: Get ShortByteString
+    let nuri = internalNormalizedFilePathToUri (wrapOsPath v)
+    return (NormalizedFilePath (fromJust nuri) v)
+
+-- Remove the CPP hack when OsPath gets 'Binary' instance.
+-- See: https://gitlab.haskell.org/haskell/filepath/-/issues/122
+
+unwrapOsPath :: OsPath -> ShortByteString
+#ifdef OS_PATH
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+unwrapOsPath = getWindowsString . getOsString
+#else
+unwrapOsPath = getPosixString . getOsString
+#endif
```

Comment:
are these exhaustive?

Reply:
I think it is. Below is how `PlatformString` defined in [`filepath`](https://hackage.haskell.org/package/filepath-1.4.100.0/docs/src/System.OsString.Internal.Types.html#OsString).
```
#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
type PlatformString = WindowsString
#else
type PlatformString = PosixString
#endif
```